Mathlib/Tactic/NormNum/NatFactorial.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2023 Sebastian Zimmer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Zimmer
+-/
+import Mathlib.Data.Nat.Factorial.Basic
+import Mathlib.Tactic.NormNum
+
+/-! # `norm_num` extensions for factorials
+
+Extensions for `norm_num` that compute `Nat.factorial`, `Nat.ascFactorial` and `Nat.descFactorial`.
+
+This is done by reducing each of these to `ascFactorial`, which is computed using a divide and
+conquer strategy that improves performance and avoids exceeding the recursion depth.
+
+-/
+
+namespace Mathlib.Meta.NormNum
+
+open Nat Qq Lean Elab.Tactic Qq Meta
+
+lemma ascFactorial_mul_ascFactorial (n l k : ℕ) :
+    n.ascFactorial l * (n + l).ascFactorial k = n.ascFactorial (l + k) := by
+  apply mul_left_cancel₀ (factorial_pos n).ne.symm
+  simp [factorial_mul_ascFactorial, ← mul_assoc, add_assoc]
```

Comment:
This lemma looks like it should live next to `factorial_mul_ascFactorial`; this is a general-purpose lemma, not just a norm-num trick.

Reply:
Agreed!